controllers/components/kserve/kserve_controller.go

```diff
@@ -19,40 +19,85 @@ package kserve
+	_, err := reconciler.ComponentReconcilerFor(mgr, componentsv1.KserveInstanceName, &componentsv1.Kserve{}).
+		// operands - owned
+		Owns(&corev1.Secret{}).
+		Owns(&corev1.Service{}).
+		Owns(&corev1.ConfigMap{}).
+		Owns(&corev1.ServiceAccount{}).
+		Owns(&rbacv1.Role{}).
+		Owns(&rbacv1.RoleBinding{}).
+		Owns(&rbacv1.ClusterRole{}).
+		Owns(&rbacv1.ClusterRoleBinding{}).
+		Owns(&templatev1.Template{}).
+		Owns(&featuresv1.FeatureTracker{}).
+		Owns(&networkingv1.NetworkPolicy{}).
+		Owns(&monitoringv1.ServiceMonitor{}).
+		Owns(&admissionregistrationv1.MutatingWebhookConfiguration{}).
+		Owns(&admissionregistrationv1.ValidatingWebhookConfiguration{}).
+		Owns(&appsv1.Deployment{}, builder.WithPredicates(resources.NewDeploymentPredicate())).
+		// operands - watched
+		//
+		// By default the Watches functions adds:
+		// - an event handler mapping to a cluster scope resource identified by the
+		//   components.opendatahub.io/managed-by annotation
+		// - a predicate that check for generation change for Delete/Updates events
+		//   for to objects that have the label components.opendatahub.io/managed-by
+		//   set to the current owner
+		//
+		Watches(&extv1.CustomResourceDefinition{}).
 
-// Reconcile is part of the main kubernetes reconciliation loop which aims to
-// move the current state of the cluster closer to the desired state.
-// TODO(user): Modify the Reconcile function to compare the state specified by
-// the Kserve object against the actual cluster state, and then
-// perform operations to make the cluster state reflect the state specified by
-// the user.
-//
-// For more details, check Reconcile and its Result here:
-// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.12.2/pkg/reconcile
-func (r *KserveReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
-
-	// TODO(user): your logic here
-
-	return ctrl.Result{}, nil
-}
+		// actions
+		WithAction(initialize).
+		WithAction(devFlags).
+		WithAction(configureServerless).
+		WithAction(configureServiceMesh).
+		WithAction(render.NewAction(
+			render.WithCache(true, render.DefaultCachingKeyFn),
+			// These are the default labels added by the legacy deploy method
+			// and should be preserved as the original plugin were affecting
+			// deployment selectors that are immutable once created, so it won't
+			// be possible to actually amend the labels in a non-disruptive
+			// manner.
+			//
+			// Additional labels/annotations MUST be added by the deploy action
+			// so they would affect only objects metadata without side effects
+			render.WithLabel(labels.ODH.Component(componentName), "true"),
+			render.WithLabel(labels.K8SCommon.PartOf, componentName),
+		)).
+		WithAction(deploy.NewAction(
+			deploy.WithFieldOwner(componentsv1.KserveInstanceName),
+			deploy.WithLabel(labels.ComponentPartOf, componentsv1.KserveInstanceName),
+		)).
+		WithAction(setupKserveConfig).
```

Comment:
i am thinking about this one................

Reply:
I didn't think about improving things muchso far, just trying to get everything to work correctly. I guess the setupKserveConfig action could be simplified so that it sets the correct config in the resources from the start, rather than deploying and then modifying.

I'm not sure how we'd handle an update case, where the default deployment mode gets changed (I guess that's legal?)...how would we know when to roll the pods vs not? Maybe we could compare the current state of the resource against what we're about to deploy, and if it will result in a different default deployment mode, then we could have a simpler post deploy action to delete the old pods? :thinking: I'll make a note to explore what I'm thinking when I've got things in better shape more generally.

Reply:
Yep I think this should be revisited. 

Not sure if it is applicable in this case, but in some of the operator I've been working on we were settle on two ways:
- generate a config map with an hash in the name (this is something also supported by kustomize itself), that would cause the deployment to restart the pods, but this has a side effect that the config map keep changing and one has to clean it up
- and an env var with an hash to the deployment, which would cause the deployment to restart the pods (this may be done by using a kustomize plugin as part of the rendering pipeline)

This avoid having to delete the PODs manually, which seems odd.

Reply:
It would be even better of this were done by the kserve itself though the kuatomize files

Reply:
> generate a config map with an hash in the name (this is something also supported by kustomize itself), that would cause the deployment to restart the pods, but this has a side effect that the config map keep changing and one has to clean it up

I think the first approach won't work here, since the configmap name is not configurable currently: it's not specified on the pod as an env var or volume, it's a constant in kserve: https://github.com/kserve/kserve/blob/21b103e1ce7166a444d0145412cee19bd6574309/python/kserve/kserve/constants/constants.py#L35

> and an env var with an hash to the deployment, which would cause the deployment to restart the pods (this may be done by using a kustomize plugin as part of the rendering pipeline)

I'll investigate this approach a little more.

Reply:
Current state of the PR adds an annotation to the pod template with a hash of the configmap, so when the defaultDeploymentMode changes, that hash will change, and the pods will be replaced. I'll close this thread, and further discussion can happen on the updated files. Thanks!